### PR TITLE
Add new lint: `rc_mutex`

### PR DESCRIPTION
**Pull request by lengyijun**
changelog: Add new lint `rc_mutex`. 

It lints on `Rc<Mutex<T>>`.  

`Rc<Mutex<T>>` should be corrected to `Rc<RefCell<T>>`

### CODE REVIEW

**Review by llogiq — CHANGES_REQUESTED** [review]
This should work well enough as a restriction lint. However, I think the suggestion is incomplete and the tests should be fleshed out more.

File: tests/ui/rc_mutex.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+pub struct SubT<T> {
+    foo: T,
+}
+
+pub enum MyEnum {
+    One,
+    Two,
+}
+
+pub fn test1<T>(foo: Rc<Mutex<T>>) {}
```
[the remark this answers is not in the document]
Comment (lengyijun) [reply tests/ui/rc_mutex.rs:21]: Could you please check if I added correct test in my latest commit?
I'm not sure I did it right.

File: tests/ui/rc_mutex.rs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+pub fn test2(foo: Rc<Mutex<MyEnum>>) {}
+
+pub fn test3(foo: Rc<Mutex<SubT<usize>>>) {}
+
+fn main() {}
```
Comment (llogiq) [inline tests/ui/rc_mutex.rs:27]: How about calling `test1`, `test2`, etc. from here?

This is something the lint might want to consider, btw., and also the reason not to apply it automatically.

**Comment by flip1995** [comment]
Is there an issue open for this lint? Or was it discussed in any other place beforehand? If so, please link to it in the PR description.

**Comment by lengyijun** [comment]
There is no related issue and discussion.

File: clippy_lints/src/types/mod.rs [hunk 3]
```diff
@@ -250,12 +251,39 @@ declare_clippy_lint! {
     "usage of very complex types that might be better factored into `type` definitions"
 }
 
+declare_clippy_lint! {
+    /// **What it does:** Checks for `Rc<Mutex<T>>`.
+    ///
+    /// **Why is this bad?** `Rc` is used in single thread and `Mutex` is used in multi thread.
+    /// Consider using `Rc<RefCell<T>>` in single thread or `Arc<Mutex<T>>` in multi thread.
+    ///
+    /// **Known problems:** None.
```
Comment (llogiq) [inline clippy_lints/src/types/mod.rs:260]: I think there may be instances where combining generics and traits could require this actual type combination, and following the suggestion would make it impossible to implement the traits.

We should consider those cases false positives and make a note of a possible problem. I don't think this problem is worrisome for the lint (it's unlikely to happen in practise), but we should note it nonetheless.

**Comment by llogiq** [comment]
Sorry for letting you wait for so long. I had some $life keeping me busy.

I'd like a more fleshed out explanation of the possible false positives. How about: "Sometimes combining generic types can lead to the requirement that a type use `Rc` in conjunction with `Mutex`. We must consider those cases false positives, but alas they are quite hard to rule out. Luckily they are also rare."

**Comment by llogiq** [comment]
Thank you! @bors r+

**Pull request merged**